src/Microsoft.CodeAnalysis.Analyzers/CSharp/MetaAnalyzers/CSharpSemanticModelGetDeclaredSymbolAlwaysReturnsNullAnalyzer.cs

```diff
@@ -0,0 +1,97 @@
+            CreateLocalizableResourceString(nameof(SemanticModelGetDeclaredSymbolAlwaysReturnsNullTitle)),
+            CreateLocalizableResourceString(nameof(SemanticModelGetDeclaredSymbolAlwaysReturnsNullForFieldMessage)),
+            DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: CreateLocalizableResourceString(nameof(SemanticModelGetDeclaredSymbolAlwaysReturnsNullForFieldDescription)),
+            helpLinkUri: null,
+            customTags: WellKnownDiagnosticTagsExtensions.Telemetry);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(static context =>
+            {
+                var typeProvider = WellKnownTypeProvider.GetOrCreate(context.Compilation);
+                IMethodSymbol? getDeclaredSymbolMethod;
+                if (!typeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisCSharpCSharpExtensions, out var csharpExtensions)
+                    || !typeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisModelExtensions, out var modelExtensions)
+                    || !typeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisCSharpSyntaxBaseFieldDeclarationSyntax, out var baseFieldDeclaration)
+                    || !typeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisSyntaxNode, out var syntaxNode)
+                    || (getDeclaredSymbolMethod = modelExtensions.GetMembers(nameof(ModelExtensions.GetDeclaredSymbol)).FirstOrDefault() as IMethodSymbol) is null)
+                {
+                    return;
+                }
+
+                var allowedTypes = csharpExtensions.GetMembers(nameof(CSharpExtensions.GetDeclaredSymbol))
+                    .OfType<IMethodSymbol>()
+                    .Where(m => m.Parameters.Length >= 2)
+                    .Select(m => m.Parameters[1].Type);
+
+                context.RegisterOperationAction(ctx => AnalyzeInvocation(ctx, getDeclaredSymbolMethod, allowedTypes, baseFieldDeclaration, syntaxNode), OperationKind.Invocation);
+            });
+        }
+
+        private static void AnalyzeInvocation(OperationAnalysisContext context, IMethodSymbol getDeclaredSymbolMethod, IEnumerable<ITypeSymbol> allowedTypes, INamedTypeSymbol baseFieldDeclarationType, INamedTypeSymbol syntaxNodeType)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            if (SymbolEqualityComparer.Default.Equals(invocation.TargetMethod, getDeclaredSymbolMethod))
+            {
+                var syntaxNodeDerivingType = invocation.Arguments[1].Value.WalkDownConversion().Type;
```

Comment:
Where do we ensure that `getDeclaredSymbolMethod` has at least 2 parameters?

Reply:
This would also not work as expected if named arguments would be used. You probably want to use one of the below helper methods:

https://github.com/dotnet/roslyn-analyzers/blob/b4ed6a3093cfd3c8d353214ce97aaa7d24cf2df1/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L850

https://github.com/dotnet/roslyn-analyzers/blob/b4ed6a3093cfd3c8d353214ce97aaa7d24cf2df1/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L871